lib/bolt/resource_instance.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+require 'bolt/target'
+require 'json'
+
+module Bolt
+  class ResourceInstance
+    attr_reader :target, :type, :title, :state, :desired_state
+    attr_accessor :events
+
+    def initialize(resource_hash)
+      @target        = resource_hash['target']
+      @type          = resource_hash['type']
+      @title         = resource_hash['title']
+      # get_resources() returns observed state under the 'parameters' key
+      @state         = resource_hash['state'] || resource_hash['parameters'] || {}
+      @desired_state = resource_hash['desired_state'] || {}
+      @events        = resource_hash['events'] || []
+
+      validate
+    end
+
+    def self.from_asserted_hash(resource_hash)
+      new(resource_hash)
+    end
+
+    def self.from_asserted_args(target,
+                                type,
+                                title,
+                                state         = nil,
+                                desired_state = nil,
+                                events        = nil)
+      new(
+        'target'        => target,
+        'type'          => type,
+        'title'         => title,
+        'state'         => state,
+        'desired_state' => desired_state,
+        'events'        => events
+      )
+    end
+
+    def validate
+      assert_type('target', target, Bolt::Target)
+      assert_type('type', type, String, Puppet::Pops::Types::PResourceType)
+      assert_type('title', title, String)
+      assert_type('state', state, Hash)
+      assert_type('desired_state', desired_state, Hash)
+      assert_type('events', events, Array)
+    end
+
+    def eql?(other)
+      self.class.equal?(other.class) &&
+        target == other.target &&
+        type   == other.type &&
+        title  == other.title
+    end
+    alias == eql?
+
+    def to_s
+      reference
+    end
```

Comment:
Should this return the reference, or something else like the resource declaration? Do we want a separate method that will return the stringified resource declaration?

```
Type { 'title': attributes }
```